core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenTaskSource.java

```diff
@@ -435,102 +407,80 @@ public Optional<List<Object>> getTableExecuteSplitsInfo()
         {
-            checkState(!started, "already started");
-            checkState(!closed, "already closed");
-            started = true;
-            processNext();
+            return delegate;
         }
 
-        private synchronized void processNext()
+        @Override
+        public void addListener(Runnable listener, Executor executor)
         {
-            if (closed) {
-                return;
+            throw new UnsupportedOperationException();
+        }
+
+        public ListenableFuture<T> addListener()
+        {
+            SettableFuture<T> listener = SettableFuture.create();
+            synchronized (listeners) {
+                listeners.add(listener);
             }
-            verify(splitLoadingFuture == null || splitLoadingFuture.isDone(), "splitLoadingFuture is still running");
-            long start = System.nanoTime();
-            splitLoadingFuture = splitSource.getNextBatch(splitBatchSize);
-            Futures.addCallback(splitLoadingFuture, new FutureCallback<>()
-            {
-                @Override
-                public void onSuccess(SplitBatch result)
-                {
-                    try {
-                        getSplitTimeRecorder.accept(start);
-                        ListMultimap<Integer, Split> splits = result.getSplits().stream()
-                                .collect(toImmutableListMultimap(splitToPartition::applyAsInt, Function.identity()));
-                        boolean finished = result.isLastBatch() && splitSource.isFinished();
-                        callback.update(splits, finished);
-                        if (!finished) {
-                            processNext();
+
+            listener.addListener(
+                    () -> {
+                        if (listener.isCancelled()) {
+                            synchronized (listeners) {
+                                listeners.remove(listener);
+                            }
                         }
-                    }
-                    catch (Throwable t) {
-                        callback.failed(t);
-                    }
-                }
+                    },
+                    directExecutor());
 
-                @Override
-                public void onFailure(Throwable t)
-                {
-                    callback.failed(t);
-                }
-            }, executor);
+            propagateIfNecessary();
+
+            return listener;
         }
 
         @Override
-        public synchronized void close()
+        public boolean cancel(boolean mayInterruptIfRunning)
         {
-            if (closed) {
-                return;
-            }
-            closed = true;
-            if (splitLoadingFuture != null) {
-                splitLoadingFuture.cancel(true);
-                splitLoadingFuture = null;
-            }
-            splitSource.close();
+            boolean result = delegate().cancel(mayInterruptIfRunning);
+            propagateIfNecessary();
```

Comment:
why do you need this if this when `propagateIfNecessary` is already attached to delegate?

Reply:
Hmm, good point. Let me remove the override